cmd/bosun/expr/azure.go

```diff
@@ -0,0 +1,532 @@
+func azureTags(arg parse.Node) (parse.Tags, error) {
+	tags := parse.Tags{"name": struct{}{}, "rsg": struct{}{}}
+	csvTags := strings.Split(arg.(*parse.StringNode).Text, ",")
+	for _, k := range csvTags {
+		tags[k] = struct{}{}
+	}
+	return tags, nil
+}
+
+// Azure API References
+// - https://docs.microsoft.com/en-us/azure/monitoring-and-diagnostics/monitoring-supported-metrics
+// - https://docs.microsoft.com/en-us/azure/monitoring-and-diagnostics/monitoring-data-sources
+
+// TODO
+// - Finish up azmd info function
+
+const azTimeFmt = "2006-01-02T15:04:05"
+
+func azResourceURI(subscription, resourceGrp, Namespace, Resource string) string {
+	return fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/%s/%s", subscription, resourceGrp, Namespace, Resource)
+}
+
+// AzureMetricDefinitions fetches metric information for a specific resource and metric tuple
+// TODO make this return and not fmt.Printf
+func AzureMetricDefinitions(prefix string, e *State, T miniprofiler.Timer, namespace, metric, rsg, resource string) (r *Results, err error) {
+	r = new(Results)
+	cc, clientFound := e.Backends.AzureMonitor[prefix]
+	if !clientFound {
+		return r, fmt.Errorf("azure client with name %v not defined", prefix)
+	}
+	c := cc.MetricDefinitionsClient
+	defs, err := c.List(context.Background(), azResourceURI(c.SubscriptionID, rsg, namespace, resource), namespace)
+	if err != nil {
+		return
+	}
+	if defs.Value == nil {
+		return r, fmt.Errorf("No metric definitions in response")
+	}
+	for _, def := range *defs.Value {
+		agtypes := []string{}
+		for _, x := range *def.SupportedAggregationTypes {
+			agtypes = append(agtypes, fmt.Sprintf("%s", x))
+		}
+		dims := []string{}
+		if def.Dimensions != nil {
+			for _, x := range *def.Dimensions {
+				dims = append(dims, fmt.Sprintf("%s", *x.Value))
+			}
+		}
+		fmt.Println(*def.Name.LocalizedValue, strings.Join(dims, ", "), strings.Join(agtypes, ", "))
+	}
+	return
+}
+
+// AzureQuery queries an Azure monitor metric for the given resource and returns a series set tagged by
+func AzureQuery(prefix string, e *State, T miniprofiler.Timer, namespace, metric, tagKeysCSV, rsg, resName, agtype, interval, sdur, edur string) (r *Results, err error) {
+	r = new(Results)
+	// Verify prefix is a defined resource and fetch the collection of clients
+	cc, clientFound := e.Backends.AzureMonitor[prefix]
+	if !clientFound {
+		return r, fmt.Errorf("azure client with name %v not defined", prefix)
+	}
+	c := cc.MetricsClient
+	r = new(Results)
+	// Parse Relative Time to absolute time
+	sd, err := opentsdb.ParseDuration(sdur)
+	if err != nil {
+		return
+	}
+	var ed opentsdb.Duration
+	if edur != "" {
+		ed, err = opentsdb.ParseDuration(edur)
+		if err != nil {
+			return
+		}
+	}
+	st := e.now.Add(time.Duration(-sd)).Format(azTimeFmt)
+	en := e.now.Add(time.Duration(-ed)).Format(azTimeFmt)
+
+	// Set Dimensions (tag) keys for metrics that support them by building a filter
```

Comment:
expand on comment , link to https://docs.microsoft.com/en-us/rest/api/monitor/filter-syntax